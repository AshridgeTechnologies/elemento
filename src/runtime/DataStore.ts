import Observable from 'zen-observable'
import lodash from 'lodash'; const {matches} = lodash;

export type Criteria = object

export type CollectionName = string
export type DataStoreObject = object
export type Id = string | number

export const InvalidateAll = 'InvalidateAll'
export const Add = 'Add'
export const Update = 'Update'
export const Remove = 'Remove'
export const MultipleChanges = 'MultipleChanges'
export type UpdateType = typeof InvalidateAll | typeof MultipleChanges | typeof Add | typeof Update | typeof Remove
export type UpdateNotification = {
    collection: CollectionName,
    type: UpdateType,
    id?: Id
    changes?: object
}
export class Pending {
    valueOf() { return null }
}

export class ErrorResult {
    constructor(public description: string, public errorMessage: string) {}
}

export function queryMatcher(criteria: Criteria) {
    return matches(criteria)
}

export interface BasicDataStore {
    getById(collection: CollectionName, id: Id): Promise<DataStoreObject | null>
    query(collection: CollectionName, criteria: Criteria): Promise<Array<DataStoreObject>>

    add(collection: CollectionName, id: Id, item: DataStoreObject): Promise<void>
    addAll(collection: CollectionName, items: {[id: Id]: DataStoreObject}): Promise<void>
    update(collection: CollectionName, id: Id, changes: object): Promise<void>
    remove(collection: CollectionName, id: Id): Promise<void>
}

export default interface DataStore extends BasicDataStore {
    observable(collection: CollectionName): Observable<UpdateNotification>
}