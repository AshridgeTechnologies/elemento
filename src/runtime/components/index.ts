export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
