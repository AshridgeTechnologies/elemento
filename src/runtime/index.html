<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>