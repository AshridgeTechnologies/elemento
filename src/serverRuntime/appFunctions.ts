import dataFunctions from '../runtime/dataFunctions'

type UserDetails = Partial<{email: string, displayName: string, password: string}>

// const CreateUser = async (uid: string, details: UserDetails) => {
//     const {email, displayName, password} = details
//     await getAuth().createUser({uid, email, password, displayName,})
// }
// const UpdateUser = async (uid: string, details: UserDetails) => {
//     const {email, displayName, password} = details
//     const changes = {email, displayName, password}
//     await getAuth().updateUser(uid, changes)
// }
// const GetUser = (uid: string) => getAuth().getUser(uid)


const appFunctions = {...dataFunctions, /*CreateUser, GetUser, UpdateUser*/ }

export default appFunctions
