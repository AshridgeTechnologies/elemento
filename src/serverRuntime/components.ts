export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts