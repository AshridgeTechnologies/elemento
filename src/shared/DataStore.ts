import Observable from 'zen-observable'
import {isPlainObject} from 'lodash';

const pendingFlag = Symbol('pending')
export type SimpleCriteria = object
export type Operator = '=' | '==' | '>' | '>=' | '<' | '<=' | '!='
export type CriteriaCondition = [fieldName: string, operator: Operator, value: any]
export type ComplexCriteria = CriteriaCondition[]
export type Criteria = SimpleCriteria | ComplexCriteria

export type CollectionName = string
export type DataStoreObject = object
export type Id = string | number

export const InvalidateAll = 'InvalidateAll'
export const Add = 'Add'
export const Update = 'Update'
export const Remove = 'Remove'
export const MultipleChanges = 'MultipleChanges'
export type UpdateType = typeof InvalidateAll | typeof MultipleChanges | typeof Add | typeof Update | typeof Remove
export type UpdateNotification = {
    collection: CollectionName,
    type: UpdateType,
    id?: Id
    changes?: object
}
export const pending = (p: Promise<any>) => {
    (p as any).valueOf = () => null;
    (p as any)[pendingFlag] = true;
    return p
}
export const isPending = (val: any) => Boolean(val?.[pendingFlag])
export class ErrorResult {
    readonly isError = true
    constructor(public description: string, public errorMessage: string) {}
    toString() {
        return `Error: ${(this.description)} - ${this.errorMessage}`
    }
    valueOf() { return null }
}

const isSimpleCriteria = (criteria: Criteria) => isPlainObject((criteria))

export function queryMatcher(criteria: Criteria) {
    const complexCriteria: ComplexCriteria = isSimpleCriteria(criteria) ? Object.entries(criteria).map(([name, value]) => [name, '=', value]) :  criteria as ComplexCriteria
    return (obj: DataStoreObject): boolean => {
        const conditionFn = (crit: CriteriaCondition) => (obj: DataStoreObject): boolean => {
            const  [fieldName, operator, value] = crit
            const fieldValue = obj[fieldName as keyof object]
            switch(operator) {
                case '=':
                case '==' : return fieldValue === value
                case '>': return fieldValue > value
                case '>=': return fieldValue >= value
                case '<': return fieldValue < value
                case '<=': return fieldValue <= value
                case '!=': return fieldValue != value
                default: return false
            }
        }
        const conditionFns = complexCriteria.map( conditionFn )
        return conditionFns.every( fn => fn(obj))
    }
}

export interface BasicDataStore {
    getById(collection: CollectionName, id: Id, nullIfNotFound: boolean): Promise<DataStoreObject | null>
    query(collection: CollectionName, criteria: Criteria): Promise<Array<DataStoreObject>>

    add(collection: CollectionName, id: Id, item: DataStoreObject): Promise<void>
    addAll(collection: CollectionName, items: {[id: Id]: DataStoreObject}): Promise<void>
    update(collection: CollectionName, id: Id, changes: object): Promise<void>
    remove(collection: CollectionName, id: Id): Promise<void>
}

export default interface DataStore extends BasicDataStore {
    observable(collection: CollectionName): Observable<UpdateNotification>
}

export const AuthStatusValues = ['readwrite', 'readonly'] as const
export type AuthStatus = typeof AuthStatusValues[number] | null
