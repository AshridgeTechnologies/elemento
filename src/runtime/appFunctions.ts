import {Value, valueOf} from './runtimeFunctions'
import {currentUser} from './components/authentication'
import dataFunctions from './dataFunctions'
import {addNotification, type NotificationLevel} from './components/notifications'
import {globalFunctions} from './globalFunctions'

const appFunctions = {

    Reset(...components: { Reset: () => void }[]) {
        components.forEach(comp => comp.Reset())
    },

    Set(component: { Set: (value: any) => void }, value: any) {
        component.Set(valueOf(value))
    },

    SetWithUpdates(component: { Set: (value: any) => void, value: object }, ...updateArgs: any[]) {
        const newValue = globalFunctions.WithUpdates(component.value, ...updateArgs)
        component.Set(newValue)
    },

    NotifyError(description: string, error: Error) {
        appFunctions.Notify('error', description, error.message)
    },

    Notify(level: Value<NotificationLevel>, message: Value<string>, details?: Value<string>) {
        addNotification(valueOf(level), valueOf(message), valueOf(details))
    },

    CurrentUser() {
        const user = currentUser()
        return user ? {...user, Name: user.displayName, Email: user.email} : null
    },

    ...dataFunctions
}

export const appFunctionsNames = () => {
    return Object.keys(appFunctions)
}

export const knownSyncAppFunctionsNames = () => ['Reset', 'Set', 'NotifyError', 'Notify', 'CurrentUser', 'GetRandomId']

export default appFunctions
