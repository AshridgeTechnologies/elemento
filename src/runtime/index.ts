import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
