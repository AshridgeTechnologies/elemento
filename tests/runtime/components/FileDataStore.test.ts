/**
 * @jest-environment jsdom
 */

import {FileDataStore, FileDataStoreImpl} from '../../../src/runtime/components/index'
import {snapshot} from '../../testutil/testHelpers'
import {createElement} from 'react'
import {render} from '@testing-library/react'

import appFunctions from '../../../src/runtime/appFunctions'
import {ErrorResult} from '../../../src/runtime/DataStore'
import Observable from 'zen-observable'

jest.mock('../../../src/runtime/appFunctions')

let dataStore: FileDataStoreImpl
let state: any

beforeEach(() => {
    dataStore = mockDataStore()
    state = new FileDataStore.State()
    state.state.dataStore = dataStore
} )

const mockObservable = new Observable(() => {})
const mockDataStore = (): FileDataStoreImpl => ({
    Open: jest.fn().mockResolvedValue(undefined),
    Save: jest.fn().mockResolvedValue(undefined),
    SaveAs: jest.fn().mockResolvedValue(undefined),
    New: jest.fn().mockResolvedValue(undefined),
    getById: jest.fn().mockResolvedValue({a:77}),
    add: jest.fn().mockResolvedValue(undefined),
    addAll: jest.fn().mockResolvedValue(undefined),
    update: jest.fn().mockResolvedValue(undefined),
    remove: jest.fn().mockResolvedValue(undefined),
    query: jest.fn().mockResolvedValue([{a:77}]),
    observable: jest.fn().mockReturnValue(mockObservable),
}) as unknown as FileDataStoreImpl

test('produces output with name',
    snapshot(createElement(FileDataStore, {path: 'app.page1.collection2', display: true}))
)

test('produces empty output with default value for display', () => {
    const {container} = render(createElement(FileDataStore, {path: 'app.page1.collection3'}))
    expect(container.innerHTML).toBe('')
})

test('creates its own data store if not supplied', () => {
    const state = new FileDataStore.State()
    const result = state.Open()
    expect(result).toBeUndefined()
})

test('delegates Open to data store', () => {
    const result = state.Open()
    expect(result).toBeUndefined()
    expect(dataStore.Open).toHaveBeenCalledWith()
})

test('delegates Save to data store', () => {
    const result = state.Save()
    expect(result).toBeUndefined()
    expect(dataStore.Save).toHaveBeenCalledWith()
})

test('delegates SaveAs to data store', () => {
    const result = state.SaveAs()
    expect(result).toBeUndefined()
    expect(dataStore.SaveAs).toHaveBeenCalledWith()
})

test('delegates New to data store', () => {
    const result = state.New()
    expect(result).toBeUndefined()
    expect(dataStore.New).toHaveBeenCalledWith()
})

test('delegates add to data store', async () => {
    const result = state.add('Widgets', 'w1', {a: 99})
    await expect(result).resolves.toBeUndefined()
    expect(dataStore.add).toHaveBeenCalledWith('Widgets', 'w1', {a: 99})
})

test('delegates addAll to data store', async () => {
    const result = state.addAll('Widgets', {w1: {a: 99}, w2: {a: 100}})
    await expect(result).resolves.toBeUndefined()
    expect(dataStore.addAll).toHaveBeenCalledWith('Widgets', {w1: {a: 99}, w2: {a: 100}})
})

test('delegates update to data store', async () => {
    const result = state.update('Widgets', 'w1', {a: 99})
    await expect(result).resolves.toBeUndefined()
    expect(dataStore.update).toHaveBeenCalledWith('Widgets', 'w1', {a: 99})
})

test('delegates remove to data store', async () => {
    const result = state.remove('Widgets', 'w1')
    await expect(result).resolves.toBeUndefined()
    expect(dataStore.remove).toHaveBeenCalledWith('Widgets', 'w1')
})

test('delegates getById to data store', async () => {
    const result = state.getById('Widgets', 'w1')
    await expect(result).resolves.toStrictEqual({a: 77})
    expect(dataStore.getById).toHaveBeenCalledWith('Widgets', 'w1')
})

test('delegates query to data store', async () => {
    const result = state.query('Widgets', {a: 33})
    await expect(result).resolves.toStrictEqual([{a: 77}])
    expect(dataStore.query).toHaveBeenCalledWith('Widgets', {a: 33})
})

test('delegates observable to data store', () => {
    const result = state.observable('Widgets')
    expect(result).toBe(mockObservable)
})

describe('handles errors', () => {
    // const NotifyError = jest.fn()
    const mock_appFunctions = appFunctions as jest.MockedObject<any>
    // mock_appFunctions.NotifyError = NotifyError

    const errorDataStore = (): FileDataStoreImpl => ({
        Open: jest.fn().mockRejectedValue(new Error('Bad open')),
        Save: jest.fn().mockRejectedValue(new Error('Bad save')),
        SaveAs: jest.fn().mockRejectedValue(new Error('Bad save as')),
        New: jest.fn().mockRejectedValue(new Error('Bad new')),
        getById: jest.fn().mockRejectedValue(new Error('Bad getById')),
        query: jest.fn().mockRejectedValue(new Error('Bad query')),
        add: jest.fn().mockRejectedValue(new Error('Bad add')),
        update: jest.fn().mockRejectedValue(new Error('Bad update')),
        remove: jest.fn().mockRejectedValue(new Error('Bad remove')),
    }) as unknown as FileDataStoreImpl

    beforeEach(() => {
        (appFunctions.NotifyError as jest.MockedFunction<any>).mockReset()
        dataStore = errorDataStore()
        state = new FileDataStore.State()
        state.state.dataStore = dataStore
    } )

    test('notifies error from Open', async () => {
        const result = await state.Open()
        expect(result).toBeUndefined()
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not open file', new Error('Bad open'))
    })

    test('notifies error from SaveAs', async () => {
        const result = await state.SaveAs()
        expect(result).toBeUndefined()
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not save to file', new Error('Bad save as'))
    })

    test('notifies error from Save', async () => {
        const result = await state.Save()
        expect(result).toBeUndefined()
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not save to file', new Error('Bad save'))
    })

    test('notifies error from New', async () => {
        const result = await state.New()
        expect(result).toBeUndefined()
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not reset to new file', new Error('Bad new'))
    })

    test('notifies error from add', async () => {
        const result = await state.add('Widgets', 'x1', {a:1})
        expect(result).toBeUndefined()
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not add item to data store', new Error('Bad add'))
    })

    test('notifies error from update', async () => {
        const result = await state.update('Widgets', 'x1', {a:1})
        expect(result).toBeUndefined()
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not update item in data store', new Error('Bad update'))
    })

    test('notifies error from remove', async () => {
        const result = await state.remove('Widgets', 'x1')
        expect(result).toBeUndefined()
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not remove item from data store', new Error('Bad remove'))
    })

    test('notifies and returns error from getById', async () => {
        const result = await state.getById('Widgets', 'x1')
        expect(result).toStrictEqual(new ErrorResult('Could not get item from data store', 'Bad getById'))
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not get item from data store', new Error('Bad getById'))
    })

    test('notifies error from query and returns empty result', async () => {
        const result = await state.query('Widgets', {a: 33})
        expect(result).toStrictEqual([])
        expect(appFunctions.NotifyError).toHaveBeenCalledWith('Could not query items in data store', new Error('Bad query'))
    })
})
