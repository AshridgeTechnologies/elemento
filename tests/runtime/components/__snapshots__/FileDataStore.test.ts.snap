// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`produces output with name 1`] = `
<div
  id="app.page1.collection2"
>
  <div>
    app.page1.collection2
  </div>
  <code
    0="F"
    1="i"
    10="S"
    11="t"
    12="o"
    13="r"
    14="e"
    15=" "
    16="a"
    17="p"
    18="p"
    19="."
    2="l"
    20="p"
    21="a"
    22="g"
    23="e"
    24="1"
    25="."
    26="c"
    27="o"
    28="l"
    29="l"
    3="e"
    30="e"
    31="c"
    32="t"
    33="i"
    34="o"
    35="n"
    36="2"
    4=" "
    5="D"
    6="a"
    7="t"
    8="a"
    9=" "
  />
</div>
`;

exports[`produces output with record values 1`] = `
<div
  id="app.page1.collection2"
>
  <div>
    app.page1.collection2
  </div>
  <code />
</div>
`;
